### ISSUE TITLE
Some developers will need to upgrade their rubygems

### ISSUE THREAD

**Issue by gravitystorm** [openstreetmap/openstreetmap-website#4364]
In #4353 we propose using a new gem, which will trigger an error if you are running rubygems < 3.4.9.

---

TLDR: if you run into an error message like:

```
/home/andy/.rbenv/versions/3.0.6/lib/ruby/3.0.0/psych.rb:457:in `parse_stream': undefined method `parse' for #<Psych::Parser:0x000055ef0ca09e38 @handler=#<Psych::Handlers::DocumentStream:0x000055ef0ca09f78
@stack=[], @last=nil, @root=nil, @start_line=nil, @start_column=nil, @end_line=nil, @end_column=nil, @block=#<Proc:0x000055ef0ca09e88 /home/andy/.rbenv/versions/3.0.6/lib/ruby/3.0.0/psych.rb:391>>,
@external_encoding=0> (NoMethodError)
```
then that's because the version of rubygems that you are using (perhaps the version that shipped with your ruby installation) is outdated and has a bug. To fix it, either upgrade to ruby 3.2+, or run:
```
gem update --system
```

---

This is going to catch a lot of developers I think, perhaps everyone who follows the installation guide. [Ubuntu 22.04 has rubygems 3.3.5](https://launchpad.net/ubuntu/jammy/+package/ruby-rubygems) for example. I did some research, and to avoid the bug you need either:

* A version of ruby that ships with psych 5+, or
* rubygems 3.4.9+

For some recent ruby releases, it ships with the following versions included:

- ruby 3.2.2 = rubygems 3.4.10, psych 5.0.1 (fine on both counts)
- ruby 3.2.1 = rubygems 3.4.6 , psych 5.0.1 (psych is new enough)
- ruby 3.1.4 = rubygems 3.3.26, psych 4.0.4 (rubygems is too old)
- ruby 3.0.6 = rubygems 3.2.33, psych 3.3.2 (rubygems is too old)

So if you are on ruby 3.2 you'll be fine and won't notice anything, but previous versions you'll run into the cryptic error message. The task here is to make the fix (i.e. run `gem update --system`) as obvious to our developers as possible.

Do you think we should add another step to the installation guide to cover this? I can't think of any other way to catch the problem before a new developer runs into it, it's going to happen before any code that we write is run, so there's no opportunity to show a better error message.

**Comment by gravitystorm** [comment]
@tomhughes writes

> > That's interesting because I couldn't reproduce it on the dev server or in a clean Ubuntu 22.04 container which both use 3.0...

Now that's also really interesting! How on earth does that work then???

**Comment by gravitystorm** [comment]
OK, I've also failed to reproduce this problem on other systems, including using our Vagrant image for Ubuntu.

My only guess is that there's some patch in the Ubuntu version of ruby or rubygems that accidentally fixes this. My dev system is Ubuntu 20.04 but using rbenv, so that might be what's different.

I'll close this for now, and if anyone else runs into this problem please describe your system and we can gather a better picture of who might be affected.

**Comment by AntonKhorev** [comment]
Ubuntu 22.04, ruby 3.0.6 installed with rbenv

```
Installing mini_racer 0.8.0 with native extensions
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /home/tony/.rbenv/versions/3.0.6/lib/ruby/gems/3.0.0/gems/mini_racer-0.8.0/ext/mini_racer_extension
/home/tony/.rbenv/versions/3.0.6/bin/ruby -I /home/tony/.rbenv/versions/3.0.6/lib/ruby/3.0.0 -r ./siteconf20231123-6495-iz864j.rb extconf.rb
checking for -lpthread... yes
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
	--with-opt-dir
	--without-opt-dir
	--with-opt-include
	--without-opt-include=${opt-dir}/include
	--with-opt-lib
	--without-opt-lib=${opt-dir}/lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=/home/tony/.rbenv/versions/3.0.6/bin/$(RUBY_BASE_NAME)
	--with-pthread-dir
	--without-pthread-dir
	--with-pthread-include
	--without-pthread-include=${pthread-dir}/include
	--with-pthread-lib
	--without-pthread-lib=${pthread-dir}/lib
	--with-pthreadlib
	--without-pthreadlib
	--enable-debug
	--disable-debug
	--enable-asan
	--disable-asan
/home/tony/.rbenv/versions/3.0.6/lib/ruby/3.0.0/psych.rb:457:in `parse_stream': undefined method `parse' for #<Psych::Parser:0x000056046bb5e238
@handler=#<Psych::Handlers::DocumentStream:0x000056046bb5e350 @stack=[], @last=nil, @root=nil, @start_line=nil, @start_column=nil, @end_line=nil, @end_column=nil, @block=#<Proc:0x000056046bb5e288
/home/tony/.rbenv/versions/3.0.6/lib/ruby/3.0.0/psych.rb:391>>, @external_encoding=0> (NoMethodError)
	from /home/tony/.rbenv/versions/3.0.6/lib/ruby/3.0.0/psych.rb:391:in `parse'
	from /home/tony/.rbenv/versions/3.0.6/lib/ruby/3.0.0/psych.rb:280:in `unsafe_load'
	from /home/tony/.rbenv/versions/3.0.6/lib/ruby/gems/3.0.0/gems/libv8-node-18.16.0.0-x86_64-linux/ext/libv8-node/location.rb:21:in `block in load!'
	from /home/tony/.rbenv/versions/3.0.6/lib/ruby/gems/3.0.0/gems/libv8-node-18.16.0.0-x86_64-linux/ext/libv8-node/location.rb:20:in `open'
	from /home/tony/.rbenv/versions/3.0.6/lib/ruby/gems/3.0.0/gems/libv8-node-18.16.0.0-x86_64-linux/ext/libv8-node/location.rb:20:in `load!'
	from /home/tony/.rbenv/versions/3.0.6/lib/ruby/gems/3.0.0/gems/libv8-node-18.16.0.0-x86_64-linux/lib/libv8/node.rb:8:in `configure_makefile'
	from extconf.rb:77:in `<main>'

To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /home/tony/.rbenv/versions/3.0.6/lib/ruby/gems/3.0.0/extensions/x86_64-linux/3.0.0/mini_racer-0.8.0/mkmf.log

extconf failed, exit code 1
```

Works fine after `gem update --system`.

I had to do something with psych before but I don't remember if it was on this machine or on another one, also Ubuntu 22.04, ruby 3.0.2, don't remember if it's system default ruby.

**Comment by dtroyer-uwhealth-org** [comment]
@gravitystorm thank you from troubleshooting internet land.
